### PR TITLE
ENH: Allow WSI reading at requested resolution + Expose Synchronize Read with VirtualReader

### PR DESCRIPTION
**Pull request by vqdang**
This PR is to pave the way for refactoring loading code for WSI inference. 
Minor changes:

- [x] Add coordinate conversion scheme via `_find_read_params_at_requested_resolution`, which mirrors `find_read_rect_params`
- [x] Adding keyword `location_at_requested=False` to `read_bounds` to explicitly denote location is at requested resolution system, not lv0 as before. Additionally, to maintain backward compatibility, default mode will always be wrt to lv0.
- [x] Modify `read_bound` for `OpenSlideWSIReader`, `OmnyxJP2WSIReader`, `VirtualWSIReader` to support above functionality
- [x] Correlation test to ensure reading return same expected region regardless of the origin baseline or reader
- [x] Coverage/Minor doc fix/Deep source fix

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -546,3 +546,82 @@ def _dir_sample_patches(_sample_patch1, _sample_patch2, tmpdir_factory):
         shutil.copy(_sample_patch2, dir_path.joinpath(_sample_patch2.basename))
 
     return dir_path
+
+
+@pytest.fixture(scope="session")
+def _sample_crc_tile(tmpdir_factory):
+    """Docstring."""
```
Comment (shaneahmed) [inline tests/conftest.py:553]: Need to add some description here

File: tiatoolbox/wsicore/wsireader.py [hunk 2]
```diff
@@ -129,13 +129,13 @@ def _relative_level_scales(self, resolution, units):
                 units.
 
         Examples:
-            >>> from tiatoolbox.wsicore.wsireader import get_wsireader
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:132]: please restore this docstring.

File: tiatoolbox/wsicore/wsireader.py [hunk 3]
```diff
@@ -129,13 +129,13 @@ def _relative_level_scales(self, resolution, units):
...
-            >>> wsi = get_wsireader(input_img="./CMU-1.ndpi")
+            >>> from tiatoolbox.wsicore import wsireader
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:132]: this should be replaced by get_wsireader

File: tiatoolbox/wsicore/wsireader.py [hunk 4]
```diff
@@ -412,9 +564,9 @@ def read_rect(
             :class:`numpy.ndarray`: array of size MxNx3 M=size[0], N=size[1]
 
         Example:
-            >>> from tiatoolbox.wsicore.wsireader import get_wsireader
+            >>> from tiatoolbox.wsicore import wsireader
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:567]: please restore get_wsireader

File: tiatoolbox/wsicore/wsireader.py [hunk 5]
```diff
@@ -412,9 +564,9 @@ def read_rect(
...
             >>> # Load a WSI image
-            >>> wsi = get_wsireader(input_img="./CMU-1.ndpi")
+            >>> wsi = wsireader.WSIReader("/path/to/a/wsi")
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:569]: please restore `get_wsireader`

File: tiatoolbox/wsicore/wsireader.py [hunk 6]
```diff
@@ -464,9 +616,9 @@ def read_rect(
 
         Examples:
 
-            >>> from tiatoolbox.wsicore.wsireader import get_wsireader
+            >>> from tiatoolbox.wsicore import wsireader
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:619]: please restore `get_wsireader`

File: tiatoolbox/wsicore/wsireader.py [hunk 7]
```diff
@@ -464,9 +616,9 @@ def read_rect(
...
             >>> # Load a WSI image
-            >>> wsi = get_wsireader(input_img="./CMU-1.ndpi")
+            >>> wsi = wsireader.WSIReader("/path/to/a/wsi")
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:621]: please restore `get_wsireader`

File: tiatoolbox/wsicore/wsireader.py (hunk 3 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/3fbfef2da40097681f72adf4f0d16bd7377f42e1

File: tiatoolbox/wsicore/wsireader.py (hunk 4 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/3fbfef2da40097681f72adf4f0d16bd7377f42e1

File: tiatoolbox/wsicore/wsireader.py (hunk 5 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/3fbfef2da40097681f72adf4f0d16bd7377f42e1

File: tiatoolbox/wsicore/wsireader.py (hunk 6 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/3fbfef2da40097681f72adf4f0d16bd7377f42e1

File: tiatoolbox/wsicore/wsireader.py (hunk 7 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/3fbfef2da40097681f72adf4f0d16bd7377f42e1

File: tests/conftest.py [hunk 8]
```diff
@@ -546,3 +546,102 @@ def _dir_sample_patches(_sample_patch1, _sample_patch2, tmpdir_factory):
...
+def _mini_wsi1_svs(tmpdir_factory):
+    """Sample pytest fixture for torch wsi dataset.
+    Download svs image for pytest.
+
+    """
+    svs_file_path = tmpdir_factory.mktemp("data").join("CMU-mini.svs")
+    if not pathlib.Path(svs_file_path).is_file():
+        print("\nDownloading %s" % svs_file_path)
+        r = requests.get(
+            "https://tiatoolbox.dcs.warwick.ac.uk/testdata/models/CMU-mini.svs"
```
Comment (shaneahmed) [inline tests/conftest.py:580]: Do we need to have a separate svs for evaluating models? Can we not use the CMU-1.svs?

File: tests/conftest.py [hunk 9]
```diff
@@ -546,3 +546,102 @@ def _dir_sample_patches(_sample_patch1, _sample_patch2, tmpdir_factory):
...
+def _mini_wsi2_jp2(tmpdir_factory):
+    """Sample pytest fixture for torch wsi dataset.
+    Download svs image for pytest.
+
+    """
+    svs_file_path = tmpdir_factory.mktemp("data").join("mini_wsi2.jp2")
+    if not pathlib.Path(svs_file_path).is_file():
+        print("\nDownloading %s" % svs_file_path)
+        r = requests.get(
+            "https://tiatoolbox.dcs.warwick.ac.uk/testdata/models/mini_wsi2.jp2"
```
Comment (shaneahmed) [inline tests/conftest.py:600]: Again the same comment, can we not use test.jp2?

File: tiatoolbox/wsicore/wsireader.py [hunk 10]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
+            resolution (int or float or tuple(float)): resolution to
+                read thumbnail at, default = 1.25 (objective power)
+            units (str): resolution units, default = "power"
+
+        Returns:
+            :py:obj:`tuple`: shape of WSI in (width, height).
+
+        Examples:
+            >>> from tiatoolbox.wsicore import wsireader
+            >>> wsi = wsireader.OpenSlideWSIReader(input_path="./CMU-1.ndpi")
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:428]: use `getwsireader`

File: tests/test_wsireader.py [hunk 11]
```diff
@@ -1,18 +1,24 @@
...
-from tiatoolbox import utils
-from tiatoolbox import cli
-from tiatoolbox.utils.exceptions import FileNotSupported
-
-import pytest
-from pytest import approx
 import pathlib
-import numpy as np
+
+# import sys
```
Comment (shaneahmed) [inline tests/test_wsireader.py:5]: remove commented import lines.

File: tests/conftest.py (hunk 8 above)
Comment (vqdang) [reply tests/conftest.py:580]: CMU is too large. Runtime will be very slow on travis. Hence why I create the mini version. Ideally, I will organize them to 2k x 2k, 4k x 4k and 8k x 8k.

File: tests/test_wsireader.py [hunk 12]
```diff
@@ -1288,11 +1374,177 @@ def test_jp2_missing_cod(_sample_jp2):
...
+        assert roi1.shape[1] == roi2.shape[1], (read_cfg, roi1.shape, roi2.shape)
+        if check_content:
+            cc = np.corrcoef(roi1[..., 0].flatten(), roi2[..., 0].flatten())
+            # this control the harshness of similarity test, how much should be?
+            assert np.min(cc) > 0.90, (cc, read_cfg, read_coord)
+
+    # * now check sync read by comparing the RoI with different base
+    # the output should be at same resolution even if source is of different base
+    msk = imread(_mini_wsi1_msk)
+    msk_reader = VirtualWSIReader(msk)
```
Comment (shaneahmed) [inline tests/test_wsireader.py:1436]: we should always be using get_wsireader unless it is a tif.

**Review by shaneahmed — COMMENTED** [review]
Thanks @vqdang 
Please can you address the comments. Once @John-P has approved we can merge this.

File: tests/conftest.py (hunk 1 above)
Comment (shaneahmed) [reply tests/conftest.py:553]: @vqdang  this docstring needs some description.

File: tests/test_wsireader.py (hunk 11 above)
Comment (vqdang) [reply tests/test_wsireader.py]: Fixed in https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/ac77bdc8bbf73150178286e8e68af148d9372cc6

File: tiatoolbox/wsicore/wsireader.py (hunk 10 above)
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py]: https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/ac77bdc8bbf73150178286e8e68af148d9372cc6

File: tests/conftest.py (hunk 9 above)
Comment (vqdang) [reply tests/conftest.py:600]: Refer back to this https://github.com/TIA-Lab/tiatoolbox/pull/123#discussion_r647232473

File: tests/test_wsireader.py (hunk 12 above)
Comment (vqdang) [reply tests/test_wsireader.py:1426]: Similar to https://github.com/TIA-Lab/tiatoolbox/blob/ac77bdc8bbf73150178286e8e68af148d9372cc6/tests/test_wsireader.py#L1174. 
These tests are specifically for VRReader so I dont think using `get_wsireader` is informative when we read the test again in the future.

File: tiatoolbox/wsicore/wsireader.py [hunk 13]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
+            resolution (float): Desired output resolution.
+            units (str): Units of scale, default = "level".
+                Supported units are:
+                - microns per pixel ('mpp')
+                - objective power ('power')
+                - pyramid / resolution level ('level')
+                - pixels per baseline pixel ('baseline')
+            precision (int, optional): Decimal places to use when
+                finding optimal scale. See
+                :func:`find_optimal_level_and_downsample` for more.
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:334]: These don't match the function signature below

**Review by John-P — CHANGES_REQUESTED** [review]

File: tiatoolbox/wsicore/wsireader.py [hunk 14]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
+            - :py:obj:`tuple` - Region size in level 0 coordinates
+                - :py:obj:`int` - Width
+                - :py:obj:`int` - Height
+            - :py:obj:`tuple` - Region location level 0 coordinates
+                - :py:obj:`int` - X location
+                - :py:obj:`int` - Y location
+        """
+        (
+            read_level,
+            read_lv_to_requested_fx
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py]: Can we make these variable naming consistent between this function, `_find_read_rect_params` and `_find_optimal_level_and_downsample` docstrings. It's a bit confusing / hard to follow otherwise.

File: tiatoolbox/wsicore/wsireader.py [hunk 15]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
+        size_at_read_lv = requested_size / read_lv_to_requested_fx
+        location_at_read_lv = requested_location / read_lv_to_requested_fx
+        output = (
+            size_at_read_lv,
+            location_at_read_lv,
+            size_at_lv0,
+            location_at_lv0,
+        )
+        # rounding up
+        output = tuple([(v + 0.5).astype(np.int64) for v in output])
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:381]: Elsewhere, when rounding I use `np.round` or python `round` which both do 'bankers' rounding. I think we should use the same rounding function to be consistent (avoid differences in when we round up or down on .5 etc). Unless there is a particular reason why it must round up. If so, then maybe `np.ceil` would be easier to read and not require a comment?

File: tiatoolbox/wsicore/wsireader.py [hunk 16]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
             post_read_scale_factor,
             baseline_read_size,
         )
+        return output
+
+    def _find_read_params_at_requested_resolution(
+            self, location, size, resolution, units):
+        """Work similar to `find_read_rect_params`.
+
+        This is similar for when location is at requested resolution.
```
[the remark this answers is not in the document]
Comment (John-P) [reply tiatoolbox/wsicore/wsireader.py]: Might also want to highlight how it is different to `_find_read_rect_params` here.

File: tiatoolbox/wsicore/wsireader.py [hunk 17]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
+        ) = self._find_optimal_level_and_downsample(
+            resolution, units,
+        )
+        info = self.info
+
+        # Do we need sanity check for input form ?
+        requested_location = np.array(location)
+        requested_size = np.array(size)
+        lv0_to_read_lv_fx = 1 / info.level_downsamples[read_level]
+        lv0_to_requested_fx = lv0_to_read_lv_fx * read_lv_to_requested_fx
```
[the remark this answers is not in the document]
Comment (John-P) [reply tiatoolbox/wsicore/wsireader.py]: Maybe something like `baseline_to_read_level` and `baseline_to_resolution` might be more clear and avoid numbers in the variable names?

File: tests/test_wsireader.py [hunk 18]
```diff
@@ -1471,13 +1721,15 @@ def test_command_line_jp2_slide_thumbnail_file_not_supported(_sample_jp2, tmp_pa
...
 
+# @pytest.mark.skip(reason="working, skip to run other test")
 def test_openslide_read_rect_edge_reflect_padding(_sample_svs):
     """Test openslide edge reflect padding for read_rect."""
     wsi = wsireader.OpenSlideWSIReader(_sample_svs)
     region = wsi.read_rect((-64, -64), (128, 128), pad_mode="reflect")
     assert 0 not in region.min(axis=-1)
 
 
+# @pytest.mark.skip(reason="working, skip to run other test")
```
Comment (mostafajahanifar) [inline tests/test_wsireader.py:1732]: isn't it better to remove all the commented `# @pytest.mark.skip(reason="working, skip to run other test")` lines before each testing function now that all tests are being performed? I guess we can avoid using this technique in the future as we can use the unit testing feature of our IDEs (like what @vqdang found on [VSCode](https://code.visualstudio.com/docs/python/testing)).

File: tiatoolbox/wsicore/wsireader.py [hunk 19]
```diff
@@ -984,43 +1128,70 @@ def read_bounds(
...
+                _,
+                post_read_scale,
+            ) = self._find_read_bounds_params(
+                bounds_at_lv0, resolution=resolution, units=units
+            )
+        else:  # duplicated portion with VirtualReader, factoring out ?
+            # Find parameters for optimal read
+            (
+                read_level,
+                bounds_at_read_lv,
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:1156]: Could we use `bounds_at_read_level` here? It's not much longer and matches the previous variable name then. Also, I think it's a bit more clear.

File: tests/test_wsireader.py (hunk 18 above)
Comment (vqdang) [reply tests/test_wsireader.py:1732]: being uncomment means it will run normally, you can check the test report to see nothing is skip.

File: tiatoolbox/wsicore/wsireader.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py:368]: `fx` means `scale_factor`, mimicking cv2.resize for scale factor (`fx`, `fy`). I have replaced with `scale_factor`

File: tests/test_wsireader.py (hunk 18 above)
Comment (vqdang) [reply tests/test_wsireader.py]: Anyways, I removed them in https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/46566bb05194e308b729a0c67e21a4d8dd993ea2

File: tiatoolbox/wsicore/wsireader.py [hunk 20]
```diff
@@ -1171,15 +1342,37 @@ def read_bounds(
         interpolation="optimise",
         pad_mode="constant",
         pad_constant_values=0,
+        location_is_at_requested=False,
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:1346]: I think this needs to be added to the parent class docstring for this function and the base class function signature updated. I think we could probably remove the word 'is' here too, e.g. `location_at_requested`.

File: tiatoolbox/wsicore/wsireader.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py]: I changed variable names in https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/8fa725a7083e00a93655578de524b49509ca5ff5 . Do they make more sense now ?

File: tiatoolbox/wsicore/wsireader.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py]: Addressed in https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/8fa725a7083e00a93655578de524b49509ca5ff5

File: tiatoolbox/wsicore/wsireader.py [hunk 21]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
+        output = (
             read_level,
             level_location,
             level_read_size,
             post_read_scale_factor,
             baseline_read_size,
         )
+        return output
+
+    def _find_read_params_at_requested_resolution(
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py]: https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/8fa725a7083e00a93655578de524b49509ca5ff5

File: tiatoolbox/wsicore/wsireader.py [hunk 22]
```diff
@@ -175,7 +175,9 @@ def level_to_downsample(x):
             resolution = 1 / resolution
         elif units == "level":
             if any(resolution >= len(info.level_downsamples)):
-                raise ValueError("Target scale level > number of levels in WSI")
+                raise ValueError(
+                    "Target scale level (%s) > number of levels (%s) in WSI"
+                    % (resolution, len(info.level_downsamples)))
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py]: Addressed in https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/8fa725a7083e00a93655578de524b49509ca5ff5

File: .gitignore [hunk 23]
```diff
@@ -110,6 +110,9 @@ examples/*.py
 # Mac generated
 .DS_Store
 
+tests/data
+tests/*.png
+tests/local_samples/
```
[the remark this answers is not in the document]
Comment (vqdang) [reply .gitignore:115]: Addressed in https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/8fa725a7083e00a93655578de524b49509ca5ff5

File: tiatoolbox/wsicore/wsireader.py [hunk 24]
```diff
@@ -301,13 +303,138 @@ def find_read_rect_params(self, location, size, resolution, units, precision=3):
...
+        """
+        (
+            read_level,
+            read_lv_to_requested_fx
+        ) = self._find_optimal_level_and_downsample(
+            resolution, units,
+        )
+        info = self.info
+
+        # Do we need sanity check for input form ?
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py:362]: Should I create an issue as tracker, or just leave this here btw ?

File: tiatoolbox/wsicore/wsireader.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py]: may be I should shorten `requested` to `req`, they become quite long now.

File: tests/conftest.py (hunk 1 above)
Comment (vqdang) [reply tests/conftest.py]: Addressed in https://github.com/TIA-Lab/tiatoolbox/pull/123/commits/16e246236228fb37ab4b215c31b3601817e46a08

**Review by John-P — DISMISSED** [review]
Could we possibly change `lv0` in variable names to `baseline` or `level_0`? I like baseline as it avoids numbers in the names and makes them look more distinct from variables with `_level` in the name e.g. `read_level`.

File: tests/conftest.py [hunk 25]
```diff
@@ -546,3 +546,102 @@ def _dir_sample_patches(_sample_patch1, _sample_patch2, tmpdir_factory):
...
+        print("\nDownloading %s" % file_path)
+        r = requests.get(
+            "https://tiatoolbox.dcs.warwick.ac.uk/testdata/models/CMU-mask.png"
+        )
+        with open(file_path, "wb") as f:
+            f.write(r.content)
+    else:
+        print("\nSkipping %s" % file_path)
+
+    return file_path
```
Comment (John-P) [inline tests/conftest.py:647]: Just noticed that these are also using the old python string interpolation. Can we use f-strings here to be consistent and I think they are recommended over the old method.

File: tests/test_wsireader.py [hunk 26]
```diff
@@ -1288,6 +1291,163 @@ def test_jp2_missing_cod(_sample_jp2):
...
+                )
+        # ~ sane check code for local debug
+        # import matplotlib.pyplot as plt
+        # plt.subplot(1, 2, 1)
+        # plt.imshow(roi1)
+        # plt.subplot(1, 2, 2)
+        # plt.imshow(roi2)
+        # plt.savefig('dump.png')
+        # plt.close()
+        # ~
```
Comment (John-P) [inline tests/test_wsireader.py:1330]: Should these be removed?

File: tiatoolbox/wsicore/wsireader.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (John-P) [reply tiatoolbox/wsicore/wsireader.py:362]: It won't hurt to add it to remind you later

File: tiatoolbox/wsicore/wsireader.py (hunk 20 above)
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py:1346]: Its already there. Unless you are referring to other place?
https://github.com/TIA-Lab/tiatoolbox/blob/6e24aec1c117736a01abc6e6b2e31defa56a2279/tiatoolbox/wsicore/wsireader.py#L712
https://github.com/TIA-Lab/tiatoolbox/blob/6e24aec1c117736a01abc6e6b2e31defa56a2279/tiatoolbox/wsicore/wsireader.py#L756

File: tests/conftest.py (hunk 25 above)
Comment (vqdang) [reply tests/conftest.py:647]: I think I may need a seperate PR to clean all of these format str. It spreads out all over the models code too.

File: tiatoolbox/wsicore/wsireader.py [hunk 27]
```diff
@@ -558,9 +728,11 @@ def read_bounds(
         :func:`read_rect`.
 
         Args:
-            bounds (tuple(int)): Tuple of (start_x, start_y, end_x,
-                end_y) i.e. (left, top, right, bottom) of the region in
-                baseline reference frame.
+            bounds (tuple(int)): By deafult, this is a tuple of (start_x,
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:731]: Typo here 'deafult' should be 'default'

File: tiatoolbox/wsicore/wsireader.py [hunk 28]
```diff
@@ -1415,39 +1621,135 @@ def read_bounds(
...
+            raise ValueError("Reference `objective_power` must not be None.")
+        mask_shape = np.array(self.img.shape[:2])[::-1]
+        mask_scale = ref_reader_info.slide_dimensions / mask_shape
+        mask_mpp = ref_reader_info.mpp * mask_scale
+        mask_obj = ref_reader_info.objective_power / mask_scale[0]
+        self._ref_info = copy.deepcopy(ref_reader_info)
+        self.info = WSIMeta(
+            file_path=self.info.file_path,
+            slide_dimensions=mask_shape,
+            level_count=1,
```
Comment (mostafajahanifar) [inline tiatoolbox/wsicore/wsireader.py:1745]: Why we are keeping only one level_count (level_dimensions, level_downsaples) here? I guess we can use all the levels information in the VirtualSlideReader here.

Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py:1745]: I'm not sure what do you mean here. VRReader always have only 1 level because it is not pyramidal (only baseline resolution). `attach_reader` just modifies its metadata such that internal reading can scale wrt the reference reader, which actually contain many levels.

File: tiatoolbox/wsicore/wsireader.py (hunk 20 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py:1346]: I agree to change it to `location_at_requested`

File: tiatoolbox/utils/visualization.py [hunk 29]
```diff
@@ -68,7 +68,7 @@ def overlay_patch_prediction(
             label_info[label_uid] = (str(label_uid), random_colour)
     else:
         # may need better error message
```
Comment (John-P) [inline tiatoolbox/utils/visualization.py:70]: May need a better error message for what? I am a little confused by this comment

File: tiatoolbox/wsicore/wsireader.py [hunk 30]
```diff
@@ -362,56 +365,57 @@ def _find_read_params_at_resolution(
         # Do we need sanity check for input form ?
         requested_location = np.array(location)
         requested_size = np.array(size)
-        lv0_to_read_lv_scale_factor = 1 / info.level_downsamples[read_level]
+        baseline_to_read_level_scale_factor = 1 / info.level_downsamples[read_level]
 
-        lv0_to_requested_lv_scale_factor = lv0_to_read_lv_scale_factor
-        lv0_to_requested_lv_scale_factor *= read_lv_to_requested_scale_factor
+        baseline_to_resolution_scale_factor = baseline_to_read_level_scale_factor
+        baseline_to_resolution_scale_factor *= read_level_to_resolution_scale_factor
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:371]: Probably not as issue but I would just be careful as `*=` can change the underlying value that is being referenced e.g. if it is a bumpy array. therefore it could be modifying `baseline_to_read_level_scale_factor` in addition to `baseline_to_resolution_scale_factor`. It might be goo to use a regular multiplication just to be safe.

File: tiatoolbox/wsicore/wsireader.py [hunk 31]
```diff
@@ -362,56 +365,57 @@ def _find_read_params_at_resolution(
...
 
-        size_at_lv0 = requested_size / lv0_to_requested_lv_scale_factor
-        location_at_lv0 = requested_location / lv0_to_requested_lv_scale_factor
-        size_at_read_lv = requested_size / read_lv_to_requested_scale_factor
-        location_at_read_lv = requested_location / read_lv_to_requested_scale_factor
+        size_at_baseline = requested_size / baseline_to_resolution_scale_factor
+        location_at_baseline = np.array(requested_location, dtype=np.float32)
+        location_at_baseline /= baseline_to_resolution_scale_factor
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:375]: Same thing here regarding `/=` and `*=` but looks like you are making a copy above with `np.array`. This could be more clear by uinsg a regular division or `.copy` as `requested_location` is already an array. Fore example:

```python
location_at_baseline = np.array(requested_location, dtype=np.float32)
location_at_baseline /= baseline_to_resolution_scale_factor
```

could simply be

```python
location_at_baseline = requested_location  / baseline_to_resolution_scale_factor
```

File: tiatoolbox/wsicore/wsireader.py [hunk 32]
```diff
@@ -709,7 +713,7 @@ def read_bounds(
         interpolation="optimise",
         pad_mode="constant",
         pad_constant_values=0,
-        location_is_at_requested=False,
+        location_at_requested=False,
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:716]: @shaneahmed  @vqdang I'm still not super keen on this kwarg name. What do you think? Just trying to make a good decision now as it's a pain to change it later. Alternative ideas:

```
baseline_coords=True
coords_space="baseline"  # Alternative being coords_space="resolution" or similar for coords at the resolution
```

File: tiatoolbox/utils/visualization.py (hunk 29 above)
Comment (vqdang) [reply tiatoolbox/utils/visualization.py:70]: https://github.com/TIA-Lab/tiatoolbox/blob/1520a389f88bc6f0b71d7f48c0e7e4446d7d97bc/tiatoolbox/utils/visualization.py#L76
https://github.com/TIA-Lab/tiatoolbox/blob/1520a389f88bc6f0b71d7f48c0e7e4446d7d97bc/tiatoolbox/utils/visualization.py#L79

For example this, will only tell sthg wrong but is not informative enough to directly point out why, making L76 and L79 very uninformative.

File: tiatoolbox/wsicore/wsireader.py (hunk 32 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py:716]: I would suggest `baseline_coords = True (default)` which will work opposite to `location_at_requested`.

File: tiatoolbox/wsicore/wsireader.py (hunk 30 above)
Comment (shaneahmed) [reply tiatoolbox/wsicore/wsireader.py:371]: I agree.

File: tiatoolbox/wsicore/wsireader.py [hunk 33]
```diff
@@ -1622,53 +1631,36 @@ def read_bounds(
...
-        location_is_at_requested=False,
+        coords_space='baseline',
         **kwargs,
     ):
 
-        # Note: In this case, `resolution` will be mapped to corresponding
-        # mpp value (We use it as basic because it has already been aligned) to
-        # allow correct scaling wrt to units being of ['baseline', 'level'].
-        # The output of `read_bounds` is therefore should match with output of
-        # the source reader using the same requested resolution.
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:1633]: I'm not really sure what you mean by this. Why is resolutions being mapped to an mpp?

Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py:1633]: This has been removed. This was when the method used in conjunction with `attach_to_reader`.

File: tiatoolbox/wsicore/wsireader.py [hunk 34]
```diff
@@ -1682,7 +1674,7 @@ def read_bounds(
             read_kwargs=kwargs,
         )
 
-        if location_is_at_requested:
+        if coords_space == 'resolution':
```
Comment (John-P) [inline tiatoolbox/wsicore/wsireader.py:1677]: Very minor thing, but here and in the function signature you use single quotes. I believe black should have changed these to double quotes? Could you double-check this file has been run through black to format it?

File: tiatoolbox/wsicore/wsireader.py [hunk 35]
```diff
@@ -522,6 +529,38 @@ def _find_tile_params(self, tile_objective_value: int):
...
+        pad_mode="constant",
+        pad_constant_values=0,
+        **kwargs,
+    ):
+        """Internal helper to perform `read_rect` at resolution.
+
+        In actuallity, `read_rect` at resolution is synonymous with
+        calling `read_bound` at resolution because `size` has always been
+        within the resolution system,
+        """
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:548]: Add a space at the end.

**Review by shaneahmed — APPROVED** [review]
Thanks @vqdang 
If John is happy with this I will merge it then.

**Review by John-P — APPROVED** [review]

**Review by mostafajahanifar — APPROVED** [review]
